### ISSUE TITLE
SSH Connection handling completely broken on Windows

### ISSUE THREAD

**Issue by mih** [datalad/datalad#2575]
It cannot open any connection. The commands that it build are fine after #2574 is fixed, but nothing works. The same command ran manually on the cmdline outside datalad works fine (even with the same paths), so it seems to be a genuine datalad issue.

**Comment by mih** [comment]
2.5 years later, trying to clone a dataset via SSH on windows (arguably a possible first point of contact with a user). The source dataset was created by `create-sibling` on a Debian server. Clone is confirmed to work with `git` on windows, and `datalad` on a different Debian box. Both times using `master` of today.

```
(base) C:\Users\mih>datalad clone mih@somehost.de:dummyds somerepo
mih@somehost.de's password: .00/1.00 [00:00<?, ? Candidate locations/s]
mm_receive_fd: no message header
mux_master_process_new_session: failed to receive fd 0 from slave
mm_receive_fd: no message header
mux_master_process_new_session: failed to receive fd 0 from slave
[ERROR] Failed to clone from any candidate source URL. Encountered errors per each url were:
| - mih@somehost.de:dummyds
  CommandError: '"git" "clone" "--progress" "mih@somehost.de:dummyds" "C:\Users\mih\somerepo"' failed with exitcode 128 [err: 'Cloning into 'C:\Users\mih\somerepo'...
mux_client_request_session: read from master failed: Connection reset by peer
Permission denied, please try again.
Permission denied, please try again.
Received disconnect from 134.94.10.217 port 22:2: Too many authentication failures
Disconnected from 134.94.10.217 port 22
CommandError: '"ssh" "-o" "ControlPath=C:\Users\mih\AppData\Local\datalad.org\datalad\Cache\sockets\24590fdb" "mih@somehost.de" "git-upload-pack 'dummyds'"' failed with exitcode 255
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.'] [install(C:\Users\mih\somerepo)]
install(error): C:\Users\mih\somerepo (dataset) [Failed to clone from any candidate source URL. Encountered errors per each url were:
- mih@somehost.de:dummyds
  CommandError: '"git" "clone" "--progress" "mih@somehost.de:dummyds" "C:\Users\mih\somerepo"' failed with exitcode 128 [err: 'Cloning into 'C:\Users\mih\somerepo'...
mux_client_request_session: read from master failed: Connection reset by peer
Permission denied, please try again.
Permission denied, please try again.
Received disconnect from 134.94.10.217 port 22:2: Too many authentication failures
Disconnected from 134.94.10.217 port 22
CommandError: '"ssh" "-o" "ControlPath=C:\Users\mih\AppData\Local\datalad.org\datalad\Cache\sockets\24590fdb" "mih@somehost.de" "git-upload-pack 'dummyds'"' failed with exitcode 255
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.']]
```

**Comment by bpoldrack** [comment]
> The same command ran manually on the cmdline outside datalad works fine (even with the same paths)

Is this still the case, too?

**Comment by mih** [comment]
> > The same command ran manually on the cmdline outside datalad works fine (even with the same paths)
> 
> Is this still the case, too?

There is no "same command" for `datalad clone`. But as stated above, I can `git clone` just fine. DataLad makes the `git clone` that it runs internally break.

**Comment by mih** [comment]
Extrapolating from https://code.visualstudio.com/docs/remote/troubleshooting#_enabling-alternate-ssh-authentication-methods the connection sharing feature (Control{Master|Path) is simply not available on windows, and will silently fail:

```
(base) C:\Users\mih>ssh -fN -o ControlMaster=auto -o ControlPersist=15m -o "ControlPath=C:\Users\mih\AppData\Local\datalad.org\datalad\Cache\sockets\d4ab8b34" somehost.de
mih@somehost.de's password:
```

This will happily create a "socket", with no error.

But it is not functional (had to enter my password to get the command to complete):

```
(base) C:\Users\mih>ssh -o "ControlPath=C:\Users\mih\AppData\Local\datalad.org\datalad\Cache\sockets\d4ab8b34" somehost.de uname
mm_receive_fd: no message header
mux_master_process_new_session: failed to receive fd 0 from slave
mux_client_request_session: read from master failed: Connection reset by peer
mih@ksomehost.de's password:
Linux
```

This implies that in order to get any SSH functionality running, the unconditional use of connection sharing has to stop.

**Comment by mih** [comment]
So best course of action seems to be to keep a (fake) control path for uniformity across platforms, but on windows not actually act on it. Have a dedicated `SSHConnection` class for Windows that does (mostly) nothing, and make `AnnexRepo._set_shared_connection()` be a no-op on windows.

**Comment by mih** [comment]
That isn't working. It leads to (yet undetermined) failed login attempts that lead to an instant ban on any server that runs fail2ban. Likely ssh calls are being made that are not able to talk to stdin. But that is just a guess.

But my feeling is that not having connection sharing is a significant break of concept that is difficult to accommodate.

**Comment by mih** [comment]
FTR: We could use `ssh -G <targethost>` to perform inspection of auth config for the target -- if only to make more meaningful error reports.

**Comment by yarikoptic** [comment]
#5042 was merged, and then into master in #5047. should be good to close
